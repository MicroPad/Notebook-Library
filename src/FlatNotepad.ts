import { Note, Notepad, Section, Trie } from './index';
import { format, parse } from 'date-fns';
import { EncryptionMethod } from './crypto';

export type FlatNotepadOptions = {
	lastModified?: Date;
	notepadAssets?: string[];
	sections?: { [internalRef: string]: FlatSection };
	notes?: { [internalRef: string]: Note };
	crypto?: EncryptionMethod;
};

export type FlatSection = {
	title: string;
	internalRef: string;
	parentRef?: string;
};

/**
 * A FlatNotepad is similar to the {@link Notepad} class, but it stores all the notes/sections
 * as in flat structures. FlatNotepads will likely be better for internal use in many situations.
 *
 * Something to remember is that all operations on this class like addSection, will return a <strong>new</strong>
 * object of this class, and not modify the existing one.
 */
export default class FlatNotepad {
	public readonly lastModified: string;
	public readonly sections: { [internalRef: string]: FlatSection };
	public readonly notes: { [internalRef: string]: Note };
	public readonly notepadAssets: string[];
	public readonly crypto?: EncryptionMethod;

	constructor(
		public readonly title: string,
		opts: FlatNotepadOptions = {}
	) {
		this.lastModified = format(opts.lastModified || new Date(), 'YYYY-MM-DDTHH:mm:ss.SSSZ');
		this.sections = opts.sections || {};
		this.notes = opts.notes || {};
		this.notepadAssets = opts.notepadAssets || [];
		if (opts.crypto) this.crypto = opts.crypto;
	}

	static makeFlatSection(title: string, parentRef?: string): FlatSection {
		const tmpSection = new Section(title);
		return { title: tmpSection.title, internalRef: tmpSection.internalRef, parentRef };
	}

	public addSection(section: FlatSection): FlatNotepad {
		return this.clone({
			sections: {
				...this.sections,
				[section.internalRef]: section
			}
		});
	}

	public addNote(note: Note): FlatNotepad {
		// Ensure our parent is just a string for the section's internalRef, not the whole Parent object
		if (typeof note.parent !== 'string') {
			note = note.clone({
				parent: (note.parent as Section).internalRef
			});
		}

		return this.clone({
			notes: {
				...this.notes,
				[note.internalRef]: note
			}
		});
	}

	public addAsset(uuid: string): FlatNotepad {
		return this.clone({
			notepadAssets: [
				...this.notepadAssets,
				uuid
			]
		});
	}

	public modified(lastModified: Date = new Date()): FlatNotepad {
		return this.clone({
			lastModified
		});
	}

	/**
	 * Unlike the {@link Notepad}, this uses an indexed lookup system. This should be faster than a
	 * traditional notepad search.
	 *
	 * @param {Trie} trie The search trie for the notepad
	 * @param {string} query Can either be a title-search or a hashtag-search
	 * @returns {Note[]}
	 */
	public search(trie: Trie, query: string): Note[] {
		return trie.search(query).map(ref => this.notes[ref]);
	}

	/**
	 * This will convert everything into the formal {@link Notepad} structure, however no {@link Asset}s will
	 * be restored. The client should rebuild the assets after this using the values in {@link notepadAssets}
	 * @returns {Notepad}
	 */
	public toNotepad(): Notepad {
		const buildSection = (flat: FlatSection): Section => {
			let section = new Section(flat.title, [], [], flat.internalRef);

			// Restore sub-sections
			Object.values(this.sections)
				.filter(s => s.parentRef === flat.internalRef)
				.map(s => section = section.addSection(buildSection(s)));

			// Restore notes
			Object.values(this.notes)
				.filter(n => n.parent === flat.internalRef)
				.map(n => section = section.addNote(n));

			return section;
		};

		let notepad = new Notepad(this.title, {
			lastModified: parse(this.lastModified),
			notepadAssets: this.notepadAssets,
			crypto: this.crypto
		});

		// Add all the sections + notes
		Object.values(this.sections)
			.filter(s => !s.parentRef)
			.forEach(s => notepad = notepad.addSection(buildSection(s)));

		return notepad;
	}

	public clone(opts: Partial<FlatNotepadOptions> = {}, title: string = this.title): FlatNotepad {
		return new FlatNotepad(title, {
			lastModified: parse(this.lastModified),
			sections: this.sections,
			notes: this.notes,
			notepadAssets: this.notepadAssets,
			...opts
		});
	}

	public pathFrom(obj: Note | FlatSection): (FlatSection | FlatNotepad)[] {
		const parents: FlatSection[] = [];

		if (!!(obj as Note).parent) { // This is a note
			obj = this.sections[(obj as Note).parent as string];
		} else {
			const parent = (obj as FlatSection).parentRef;
			if (!parent) return [this];

			obj = this.sections[(obj as FlatSection).parentRef!];
		}


		let tmp: FlatSection = obj;
		while (true) {
			parents.unshift(tmp);

			if (!tmp.parentRef) return [ this, ...parents ];
			tmp = this.sections[tmp.parentRef];
		}
	}
}