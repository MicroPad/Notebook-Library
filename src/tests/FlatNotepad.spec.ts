import { FlatNotepad, Note, Section } from '../index';
import { FlatNotepadOptions, FlatSection } from '../FlatNotepad';
import { TestUtils } from './TestUtils';

describe('FlatNotepad', () => {
	let options = getOptions();

	beforeEach(() => {
		options = getOptions();
	});

	describe('constructor', () => {
		it('should construct with just a title', () => {
			// Arrange
			const title = 'test';

			// Act
			const n = new FlatNotepad(title);

			// Assert
			expect(n).toBeInstanceOf(FlatNotepad);
			expect(n.title).toEqual(title);
		});

		Object.entries(options).forEach(option =>
			it(`should construct with ${option[0]}`, () => {
				// Arrange
				const title = 'test';

				// Act
				const n = new FlatNotepad(title, {
					[option[0]]: option[1]
				});

				// Assert
				expect(n).toBeInstanceOf(FlatNotepad);
				expect(n.title).toEqual(title);
				expect(n[option[0]]).toMatchSnapshot();
			})
		);
	});

	describe('addSection', () => {
		let notepad: FlatNotepad;
		let section: FlatSection;

		beforeEach(() => {
			notepad = new FlatNotepad('test');
			section = { title: 'test', internalRef: 'abc' };
		});

		it('should add a new section', () => {
			//Arrange
			// Act
			const res = notepad.addSection(section);

			// Assert
			expect(res.sections[section.internalRef]).toEqual(section);
		});

		it('should create a new object', () => {
			//Arrange
			// Act
			const res = notepad.addSection(section);

			// Assert
			expect(res).not.toBe(notepad);
		});
	});

	describe('addNote', () => {
		let notepad: FlatNotepad;
		let section: FlatSection;
		let note: Note;

		beforeEach(() => {
			notepad = new FlatNotepad('test');
			section = { title: 'test', internalRef: 'abc' };

			note = new Note('test', new Date(1).getTime()).clone({ internalRef: 'etc' });
			note.parent = 'abc';
		});

		it('should add a new note', () => {
			//Arrange
			// Act
			const res = notepad.addNote(note);

			// Assert
			expect(res.notes[note.internalRef]).toEqual(note);
		});

		it(`should add a new note and convert the parent if it isn't a string`, () => {
			//Arrange
			note.parent = new Section('test').clone({ internalRef: 'abc' });

			// Act
			const res = notepad.addNote(note);

			// Assert
			expect(res.notes[note.internalRef]).toEqual(note);
			expect(res.notes[note.internalRef].parent).toEqual('abc');
		});

		it('should create a new object', () => {
			//Arrange
			// Act
			const res = notepad.addNote(note);

			// Assert
			expect(res).not.toBe(notepad);
		});
	});

	describe('addAsset', () => {
		let notepad: FlatNotepad;
		const asset = 'test';

		beforeEach(() => {
			notepad = new FlatNotepad('test');
		});

		it('should add a new asset', () => {
			//Arrange
			// Act
			const res = notepad.addAsset(asset);

			// Assert
			expect(res.notepadAssets[0]).toEqual(asset);
		});

		it('should create a new object', () => {
			//Arrange
			// Act
			const res = notepad.addAsset(asset);

			// Assert
			expect(res).not.toBe(notepad);
		});
	});

	describe('modified', () => {
		it('should update lastModified', () => {
			// Arrange
			const expected = new FlatNotepad('expected', { lastModified: new Date(32) });
			let notepad = new FlatNotepad('test');

			// Act
			notepad = notepad.modified(new Date(32));

			// Assert
			expect(notepad.lastModified).toEqual(expected.lastModified);
		});
	});

	describe('search', () => {
		it('should call search on all notes', () => {
			// Arrange
			const notepad = new FlatNotepad('test', {
				lastModified: new Date(1),
				notes: {
					abc: TestUtils.makeNote('hi')
				}
			});

			Note.prototype.search = jest.fn(() => [TestUtils.makeNote('hi')]);

			// Act
			notepad.search('h');

			// Assert
			expect(Note.prototype.search).toHaveBeenCalledWith('h');
		});
	});

	describe('toNotepad', () => {
		it('should convert to a full Notepad object', () => {
			// Arrange
			let notepad = new FlatNotepad('test', options);
			notepad = notepad.addSection({ title: 'one-deep', internalRef: '1d', parentRef: 'abc' });
			notepad = notepad.addSection({ title: 'another root one', internalRef: 'r' });

			// Act
			const res = notepad.toNotepad();

			// Assert
			expect(res).toMatchSnapshot();
		});
	});
});

function getOptions(): FlatNotepadOptions {
	const testSection: FlatSection = { title: 'test section', internalRef: 'abc' };
	const testNote = new Note('test note', new Date(1).getTime()).clone({ internalRef: 'etc' });
	testNote.parent = 'abc';

	return {
		lastModified: new Date(1),
		sections: { abc: testSection },
		notes: { etc: testNote },
		notepadAssets: ['test']
	};
}
