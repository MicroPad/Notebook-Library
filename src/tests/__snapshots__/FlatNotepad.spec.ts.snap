// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlatNotepad constructor should construct with lastModified 1`] = `"1970-01-01T12:00:00.001+12:00"`;

exports[`FlatNotepad constructor should construct with notepadAssets 1`] = `
Array [
  "test",
]
`;

exports[`FlatNotepad constructor should construct with notes 1`] = `
Object {
  "etc": Note {
    "bibliography": Array [],
    "elements": Array [],
    "internalRef": "etc",
    "parent": "abc",
    "time": 1,
    "title": "test note",
  },
}
`;

exports[`FlatNotepad constructor should construct with sections 1`] = `
Object {
  "abc": Object {
    "internalRef": "abc",
    "title": "test section",
  },
}
`;

exports[`FlatNotepad toNotepad should convert to a full Notepad object 1`] = `
Notepad {
  "assets": Array [],
  "lastModified": "1970-01-01T12:00:00.001+12:00",
  "notepadAssets": Array [
    "test",
  ],
  "sections": Array [
    Section {
      "internalRef": "abc",
      "notes": Array [
        Note {
          "bibliography": Array [],
          "elements": Array [],
          "internalRef": "etc",
          "parent": [Circular],
          "time": 1,
          "title": "test note",
        },
      ],
      "parent": Notepad {
        "assets": Array [],
        "lastModified": "1970-01-01T12:00:00.001+12:00",
        "notepadAssets": Array [
          "test",
        ],
        "sections": Array [
          [Circular],
        ],
        "title": "test",
      },
      "sections": Array [
        Section {
          "internalRef": "1d",
          "notes": Array [],
          "parent": Section {
            "internalRef": "abc",
            "notes": Array [],
            "sections": [Circular],
            "title": "test section",
          },
          "sections": Array [],
          "title": "one-deep",
        },
      ],
      "title": "test section",
    },
    Section {
      "internalRef": "r",
      "notes": Array [],
      "parent": [Circular],
      "sections": Array [],
      "title": "another root one",
    },
  ],
  "title": "test",
}
`;
